lib/Sema/TypeCheckStmt.cpp

```diff
@@ -1246,8 +1246,21 @@ Stmt *StmtChecker::visitBraceStmt(BraceStmt *BS) {
 
       bool hadTypeError = TC.typeCheckExpression(SubExpr, DC, TypeLoc(),
                                                  CTP_Unused, options);
-      if (isDiscarded && !hadTypeError)
+
+      // If a closure expression is unused, the user might have intended
+      // to write "do { ... }".
+      auto *CE = dyn_cast<ClosureExpr>(SubExpr);
```

Comment:
Oh good point; if we're already in a BraceStmt and the immediate child is a closure, it must be "unused". I hadn't actually read the isDiscarded definition above.